eventAuditor: pattern-map, process-spec-map and process-filter-map management (userspace)

geyslan opened the pull request:
- [x] `Makefile` for compilation of a static `libbpf.a`
- [x] Create generic shared maps management logic
    - [x] Map loading/pinning (initialization)
    - [x] Map unloading/unpinning
- [x] Convert already commited code to use [kubearmor/libbpf](https://github.com/kubearmor/libbpf)
- [x] `pattern-map`, `process-spec-map` and `process-filter-map` tasks
  - [x] Implement the skeleton code for pattern-map (introduced by the last draft)
  - [x] Update pattern to the pattern-map from the userspace (introduced by the last draft)
  - [x] Implement the skeleton code for pattern-map, process-spec-map and process-filter-map management
  - [x] Update process spec to the process-spec-map from the userspace

Commit:
eventAuditor: add libbpf submodule

Commit:
eventAuditor: Makefile logic

Commit:
eventAuditor: add Feeder to EventAuditor

Commit:
eventAuditor: add InitShareMaps() logic and update dependencies

nam-jaehyun commented on KubeArmor/eventAuditor/eventAuditor.go:
```diff
@@ -1,12 +1,14 @@
 // EventAuditor Structure
 type EventAuditor struct {
-	//
+	LogFeeder *fd.Feeder
```
we will eventually use the log feeder later, but we don't need to think about it at this point.

geyslan replied:
I see. I added it once we must stick to an error checking and logging logic, like the comment below. So I'll maintain this change for now. That's ok?

geyslan commented on KubeArmor/eventAuditor/sharedMapManager.go:
```diff
@@ -1,8 +1,79 @@
+		if err != nil {
+			ea.LogFeeder.Printf(err.Error())
+			continue
+		}
```
Error Checking and Logging Logic.

It would be changed to:

```go
		if err != nil {
			ea.LogFeeder.Errf(err.Error())
			continue
		}
```
@nam-jaehyun, what do you think?

Commit:
eventAuditor: add function comments

Commit:
eventAuditor: fix var naming typo, return error properly

Commit:
eventAuditor: add functions, add struct and refactor

geyslan commented on KubeArmor/eventAuditor/sharedMapManager.go:
```diff
@@ -1,8 +1,146 @@
+	eaMap *lbpf.BPFMap
+}
+
+var sharedMods map[string]*eventAuditorEBPFModule
```
@nam-jaehyun 

BPFMap has a private pointer for the Module that contains it, but libbpfgo doesn't provide it for the consumer of the BPFMap. The solution that I found was to maintain the Map and its respective Module on our eventAuditorEBPFModule.

Why am I doing this? Well I think it's better to have an unique module for every single pinned map. This would make maintenance easier.

geyslan replied:
This workaround won't be necessary anymore. The https://github.com/aquasecurity/libbpfgo/pull/41 solved this question.

Commit:
eventAuditor: initialize map variables, improve error checking

Commit:
eventAuditor: reduce references on error checking

Even though the length of unstacked variables needs to be accessed indirectly
more than once, it's better to have fewer references to change in a possible
refactoring.

Commit:
eventAuditor: fix dependencies, update go.mod and go.sum

Commit:
eventAuditor: refactor based on libbpfgo new features

The PR https://github.com/aquasecurity/libbpfgo/pull/41 has added new features
to libbpfgo.

Due to one of these new features GetModule(), it's no longer necessary to store
pointers to BPFModule.

This also reduces code for pinMap() and unpinMap() making use of the new
GetName() BPFMap method and using the error logic from libbpfgo.

nyrahul commented on KubeArmor/eventAuditor/BPF/Makefile:
```diff
@@ -0,0 +1,92 @@
+ifeq ($(V), 1)
```
If this Makefile is inspired by some existing Makefile/repo, please ensure we attribute that repo/project in the header. For e.g, 
```
This makefile is inspired by libbpf/bootstrap.
```
If there is license attribution needed, please add that too.

geyslan replied:
@nyrahul I don't really know how to determine the composition and creation of this Makefile. Let's see. Here we are using a libbpf submodule that needs to be compiled beforehand, so the Makefile will require rules for that. For the correct use of libbpf we'll need vmlinux.h, so we have to use bpftool and create some rules to create it and check the btf kernel file. Later, we'll need to compile the bpf objects that also require other rules. 

Has libbpf-bootstrap some similar steps? Yes it has (and it does have license).

https://github.com/libbpf/libbpf-bootstrap/blob/master/examples/c/Makefile

Has libbpfgo some similar steps? Yes it has (and it doesn't have license).

https://github.com/aquasecurity/libbpfgo/blob/main/Makefile

Has linux/tools/testing/selftests/bpf some similar steps? Yes it has (and it does have license).

https://github.com/torvalds/linux/blob/master/tools/testing/selftests/bpf/Makefile

I suppose that the steps **needed** to build the solution are not attached to the solution license, but let me know. Thanks.

nyrahul replied:
The logic is plain and simple. If you did copy-paste verbatim from some existing Makefile then it is always best to attribute the source (whichever source you picked up from). It doesn't hurt to say, this Makefile was inspired from so and so Makefile.

geyslan replied:
> It doesn't hurt to say, this Makefile was inspired from so and so Makefile.

Exactly. I totally agree. I just wasn't certain since this file was not copied-pasted verbatim, but inspired from various Makefiles (edit: and written from scratch). However, as you suggested I'll include the information about the inspirations and the related licenses to avoid further implications.

geyslan replied:
Done.

nyrahul reviewed (changes requested):
Attribution and license header check.

Commit:
eventAuditor: license attribution

Commit:
eventAuditor: license and inspiration attribution

Commit:
eventAuditor: update go mod to use kubearmor/libbpf

Commit:
eventAuditor: change object file folder and add existence check

We can't just rely on the library verification.

Commit:
eventAuditor: update ebpf map names

Commit:
eventAuditor: update ebpf management code to use kubeamor/libbpf

Commit:
eventAuditor: add .gitignore to eventAuditor subdirectory

Commit:
eventAuditor: update Makefile regarding bpf object building

This changes the intermediary folder name from 'output' to 'include'.
This also introduces the ebpf 'objs' folder, where the .d and .o files are created.

Commit:
eventAuditor: add ebpf source files (pattern-map, process-spec-map)

Commit:
eventAuditor: add newline at end of files

Commit:
eventAuditor: fix maps name

Commit:
eventAuditor: add process-filter-map ebpf code

Commit:
eventAuditor: add pattern-map to sharedMapManager

Commit:
eventAuditor: update go module to the new libbpf

Commit:
eventAuditor: update BPF files (naming, structure, dependencies)

Commit:
eventAuditor: introduce sharedMaps.go and more

This adds the sharedMaps.go which contains the maps structures andc
management logic.

This also udates sharedMapManager.go to use the latest kubearmor/libbpf features
at the same time doing some refactoring and cleaning.

Commit:
eventAuditor: add tests/

Commit:
eventAuditor: go mod tidy

Commit:
eventAuditor: fix and update Makefile for tests/

Commit:
eventAuditor: add license, does some renaming and cleanup

Commit:
eventAuditor: add comments and fix function names

Commit:
eventAuditor: change logfeeder output to stdout and git rm output empty file
